bugsnag-plugin-android-anr/src/main/java/com/bugsnag/android/AnrPlugin.kt

```diff
@@ -9,31 +9,56 @@ internal class AnrPlugin : Plugin {
         private const val LOAD_ERR_MSG = "Native library could not be linked. Bugsnag will " +
             "not report ANRs. See https://docs.bugsnag.com/platforms/android/anr-link-errors"
 
-        /**
-         * Returns the index of the JVM frame that led to the native ANR, or -1 if not found.
-         */
-        internal fun getNativeANRIndex(stackTrace: Array<StackTraceElement>): Int {
-            // This is not perfect, because some JVM calls are technically native (like sleep).
-            val notifyIndex = stackTrace.indexOfFirst { frame ->
-                frame.methodName == "notifyAnrDetected" && frame.className == "com.bugsnag.android.AnrPlugin"
-            }
-            if (notifyIndex < 0) {
-                return -1
+        internal fun hasReservedNativeFrames(stackTrace: List<NativeStackframe>): Boolean {
+            val index = stackTrace.indexOfFirst { frame ->
+                val method = frame.method
+                method != null && method.startsWith("_Z")
             }
+            return index >= 0
+        }
 
-            val nativeAnrIndex = notifyIndex + 1
-            if (nativeAnrIndex >= stackTrace.size || !stackTrace[nativeAnrIndex].isNativeMethod) {
-                return -1
-            }
+        internal fun isFirstNativeFrameBlank(stackTrace: List<NativeStackframe>): Boolean {
+            val firstFrame = stackTrace.last()
+            val method = firstFrame.method
+            val file = firstFrame.file
+            val lineNumber = firstFrame.lineNumber
+            val symbolAddress = firstFrame.symbolAddress
+            val loadAddress = firstFrame.loadAddress
+
+            return (method == null || method.isEmpty()) &&
+                (file == null || file.isEmpty()) &&
+                (lineNumber == null || lineNumber == 0) &&
+                (symbolAddress == null || symbolAddress == 0L) &&
+                (loadAddress == null || loadAddress == 0L)
+        }
 
-            return nativeAnrIndex
+        internal fun isRealNativeTrace(nativeTrace: List<NativeStackframe>): Boolean {
+            // Native trace heuristics are tricky. The JVM is always reached via a native frame
+            // of some sort, so at least one frame of the Java stack trace will always return true
+            // in response to isNativeMethod().
+            //
+            // There are two known telltale signs in a native trace to indicate that only Java code
+            // is involved and the native trace is a red herring:
+            //
+            // - If it's in 100% Java land, the first native frame is "empty" in that every field
+            //   is blank except for frame address.
+            //
+            // - If it's in a system-generated native bridge such as from Thread.sleep(), the trace
+            //   will contain reserved methods starting with "_Z", for example
+            //   "_ZN3art17ConditionVariable9TimedWaitEPNS_6ThreadEli"
+            //
```

Comment:
Actually, I think I may have a better way to detect real native traces... Stay tuned!